File: m/pca_example.m
n = 125; % number of observations
m = 2;  % number of attributes (dimensions)

% generate normal samples
% X = zeros([n m]);
noise1 = randn([n 1]);
noise2 = randn([n 1]);
X = 2*noise1 + 5;
Y = .5*X + .5*noise2 + 3;
plot(X, Y, '.')
axis([0 10 0 10])
axis square

[coeff,score,latent,~,explained,mu] = pca([X Y]);

X0 = X - mu(1);
Y0 = Y - mu(2);
m = sum(X0 .* Y0) / sum(X0.^2);

%%

File: py/pca-example.py
import numpy as np
from pca import *

# Random number generator for repeatability
rng = np.random.default_rng(12)
# Set data dimensions
d = 3  # number of columns (variables)
n = 30 # number of rows (observations)
# Generate uncorrelated and correlated data
A = rng.standard_normal((n,d)) # uncorrelated
B = A.copy() # correlated
B[:,0] += 4*B[:,1] + 2*B[:,2]  # correlated
# PCA projections
coeff, score, latent, mu = pca(A, n_components=2)
Ahat = score @ coeff.T + mu # reconstruct uncorrelated data
coeff, score, latent, mu = pca(B, n_components=2)
Bhat = score @ coeff.T + mu # reconstruct correlated data
# Reconstruction error
err_uncorr = np.linalg.norm(A - Ahat)
err_corr = np.linalg.norm(B - Bhat)
print(f"Uncorrelated reconstruction error: {err_uncorr}")
print(f"Correlated reconstruction error: {err_corr}")
# Plot Scatter
pairwise3dscatter(A, Ahat, "Uncorrelated Data")
pairwise3dscatter(B, Bhat, "Correlated Data")